driver-core/src/main/com/mongodb/client/model/bulk/ClientBulkWriteOptions.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.client.model.bulk;
+
+import com.mongodb.annotations.Sealed;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.result.bulk.ClientBulkWriteResult;
+import com.mongodb.internal.client.model.bulk.ConcreteClientBulkWriteOptions;
+import com.mongodb.lang.Nullable;
+import org.bson.BsonValue;
+import org.bson.conversions.Bson;
+
+/**
+ * The options to apply when executing a client-level bulk write operation.
+ *
+ * @since 5.3
+ */
+@Sealed
+public interface ClientBulkWriteOptions {
```

Comment:
Is it worth the inconsistency with the rest of our top level `*Options` types to make this an interface?

I note that we have followed the interface pattern for other Options types, but only on the ones that extend `Bson`, not on ones that are passed directly to CRUD methods.

Reply:
Is it the way we construct `ClientBulkWriteOptions` (`clientBulkWriteOptions()` vs `new BulkWriteOptions()`) or the mere fact that it's an interface and not a concrete class, that you see as a possible issue?

1. If it's the former, then all other options related to the bulk write API, as well as write models are also created differently: `clientUpdateOptions()` vs `new UpdateOptions()`, `updateOne(...)` vs `new UpdateOneModel()`. Do you find these differences concerning?
2. If it's the latter, then `ClientWriteModel` is an interface, while `WriteModel` is an abstract class. Is this an issue?

(the above are questions, not arguments for or against something)

Exposing concrete classes, like when we expect applications to call `new BulkWriteOptions()`, gives us less flexibility (though I am not sure why would we need it for options), and almost forces us to expose the way we consume `BulkWriteOptions` (either getter methods, or a different method). It is still possible to not expose it, but requires introducing another (internal) options class, and copy the data from `BulkWriteOptions`, which is ugly.

Reply:
More the former than the latter.  And yes, I find all the differences concerning, so I want to make sure they are well motivated and worth the inconsistency.  

I don't see why we need the flexibility of interfaces for simple data-bearing classes describing options. 

And even if we leave it as an interface, it's not clear to me why the interface doesn't contain getters so that both driver and applications can examine the options without resorting to a cast to an internal type (I can imagine applications inadvertently depending on the internal type in order to examine an options instance that was passed to it by some higher level code).

Reply:
> I don't see why we need the flexibility of interfaces for simple data-bearing classes describing options.

All of a sudden, we have a case for it: https://mongodb.slack.com/archives/C035ZJL6CQN/p1722366764071179?thread_ts=1722308615.042579&cid=C035ZJL6CQN:

> We don't have `batchSize` in the public API for bulk write options so drivers would need to support it as an internal option for testing, but we could probably make that work

When we expose options as an interface and use an internal mutable implementation, we can easily add a internal option on that implementation without exposing it to the API. This seem to be a better approach than introducing another options class to represent internal options, and than trying to find a way to attach that additional object to somehow to the `bulkWrite` operation "request" that comes from the API.

Reply:
I don't agree that our API should be abused for this purpose :)

Since we'll have to specially handle such a private option anyway, we can do anything we want, including setting some private global state (like we do for testing logging iirc).

Reply:
But that's the point, it's not the API that will be used for this, but its implementation. And the place where it will be possible (`ConcreteClientBulkWriteOptions`) makes much more sense than any other place.

Reply:
I understand.  But still I don't think we should let our API choices be influenced by the need to pass private options like batchSize in.

Reply:
One more argument for modelling our inputs such that we can either hide parts of their implementation or all of the implementation: instead of wrapping `ClientInsertOneModel`s into something akin to `WriteRequestWithIndex`, and producing more garbage, we can just have a the `index` field on the internal implementation of `ClientInsertOneModel`.

Reply:
A general rule of thumb has always been the driver should feel like its written by a single person.  The idea is that the common patterns are repeated to lower the cognitive cost for users and to reduce any irksome-ness of using the driver.

Could things be done differently / better often has a different answer than _**should**_ a different pattern be used.

I think when the burden of historical design choices becomes to high, then its time for a global API change / redesign. Which essentially would be a new driver.

Reply:
General rules aside we have 3 different ways of setting options in the driver:

- Mutable class - since 3.x
- Builder class - since 3.x
- Interface and internal implementation eg: `FillOptions.fillOptions()` - since 4.7

So Interface and concrete hidden approach appears to be the pattern to use for new options classes in the modern parts of the codebase.